### PR TITLE
Idefics: enable generation tests

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Enables GenerationTesterMixin for Idefics models. Tests are passing for me locally.

There are a few PRs with idefics already and they need to be reviewed/merged in the following order: https://github.com/huggingface/transformers/pull/33907 -> https://github.com/huggingface/transformers/pull/34043 -> current PR

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -4261,7 +4262,8 @@ def _assisted_decoding(
                             added_len,
                             is_decoder_attention=True,
                         )
-                    else:
+                    # some (V)LLMs have hard requirement on SDPA and thus never return attn
+                    elif outputs.attentions[0] is not None:
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:4266]: ouuu very nice!

**Review by ArthurZucker — APPROVED** [review]
LGTM, thanks for debugging these tests!

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -1567,13 +1572,13 @@ def test_generate_from_inputs_embeds_decoder_only(self):
             # This test is for decoder-only models (encoder-decoder models have native input embeddings support in the
             # decoder)
             if config.is_encoder_decoder:
-                continue
+                self.skipTest("This test is for decoder-only models")
```
Comment (gante) [inline tests/generation/test_utils.py:1575]: We need a continue here :P

With a skip, we skip for all other models in `self.all_generative_model_classes`, and we might have encoder-decoder and decoder-only models mixed up together in that iterator

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -1567,13 +1572,13 @@ def test_generate_from_inputs_embeds_decoder_only(self):
...
 
             # Skip models without explicit support
             config.is_decoder = True
             model = model_class(config).to(torch_device).eval()
             if "inputs_embeds" not in inspect.signature(model.prepare_inputs_for_generation).parameters.keys():
-                continue
+                self.skipTest("This model doesn't accept `inputs_embeds` in generate")
```
Comment (gante) [inline tests/generation/test_utils.py:1581]: Likewise, this also needs to be a continue

**Review by gante — APPROVED** [review]
LGTM, except for those two `continue` replacements.

(tests 💛 )

File: tests/generation/test_utils.py (hunk 2 above)
Comment (zucchini-nlp) [reply tests/generation/test_utils.py:1575]: hmm oke, didn't think we could have cases with mixed model types. Was just confused the test was passing for one model while not other, and took me a while to figure out hehe

**Pull request merged**